### PR TITLE
feat(menu,refinementList): add header/item/footer templating solution

### PR DESCRIPTION
**Pull request by vvo**
BREAKING CHANGES:

Removed from menu and refinementList:
- rootClass => cssClasses.root
- itemCLass => cssClasses.item
- template => templates.item

Added to menu and refinementList:
- cssClasses{root,list,item}
- templates{header,item,footer}
- widget (container) is automatically hidden by default
- hideWhenNoResults=true

This was done to allow more templating solutions like discussed in #101.

### CODE REVIEW

**Comment by vvo** [comment]
If you are ok with the new templates/cssClasses I will update the relevant documentation.

Then I will PR again to normalize every refinement widget to include templates/cssClasses like refinementList and menu does.

cc @maxiloc @Jerskouille

File: components/RefinementList.js [hunk 1]
```diff
@@ -38,37 +39,68 @@ class RefinementList extends React.Component {
...
-    var template = this.props.template;
+    var templates = this.props.templates;
+    var rootClass = cx(this.props.cssClasses.root);
+    var listClass = cx(this.props.cssClasses.list);
+    var itemClass = cx(this.props.cssClasses.item);
 
     return (
-      <div className={this.props.rootClass}>
+      <div className={rootClass}>
+        <Template template={templates.header} />
```
Comment (redox) [inline components/RefinementList.js]: Can we ensure passing `null` or `undefined` is OK here?

**Comment by maxiloc** [comment]
:+1: Good for me

File: components/RefinementList.js (hunk 1 above)
Comment (vvo) [inline components/RefinementList.js]: Yes I will do

**Comment by Jerskouille** [comment]
:+1: Perfect!

File: components/RefinementList.js [hunk 2]
```diff
@@ -38,37 +39,68 @@ class RefinementList extends React.Component {
...
 RefinementList.propTypes = {
-  rootClass: React.PropTypes.oneOfType([
-    React.PropTypes.string,
-    React.PropTypes.arrayOf(React.PropTypes.string)
-  ]),
-  itemClass: React.PropTypes.oneOfType([
-    React.PropTypes.string,
-    React.PropTypes.arrayOf(React.PropTypes.string)
-  ]),
+  cssClasses: React.PropTypes.shape({
```
Comment (pixelastic) [inline components/RefinementList.js]: Can we define our own `PropTypes` to have a `PropTypes.cssClasses` and `PropTypes.templates` ?

File: lib/style.css [hunk 3]
```diff
@@ -0,0 +1,3 @@
+.as-display-none {
```
Comment (pixelastic) [inline lib/style.css]: What about `.as-hidden` instead?

File: widgets/refinement-list.js [hunk 4]
```diff
@@ -51,17 +57,30 @@ function refinementList({
     throw new Error(usage);
   }
 
+  if (templates !== defaultTemplates) {
+    templates = Object.assign({}, defaultTemplates, templates);
```
Comment (pixelastic) [inline widgets/refinement-list.js]: What is the `if` for? Can't we always `Object.assign({}, defaultTemplates, templates);`?

File: widgets/refinement-list.js [hunk 5]
```diff
@@ -51,17 +57,30 @@ function refinementList({
...
+  }
+
   return {
     getConfiguration: () => ({
       [operator === 'and' ? 'facets' : 'disjunctiveFacets']: [facetName]
     }),
     render: function({results, helper}) {
+      var values = results.getFacetValues(facetName, {sortBy: sortBy}).slice(0, limit);
+
+      if (values.length === 0) {
```
Comment (pixelastic) [inline widgets/refinement-list.js]: Couldn't this copy/pasted block goes into `utils` under an `hideContainerNode(containerNode, values.length === 0)`? Ie. hide the first argument if the second is true.

File: components/RefinementList.js (hunk 2 above)
Comment (vvo) [inline components/RefinementList.js]: Yes I think we could have them stored somewhere in a util (util.propTypes..). Is that it?

File: lib/style.css (hunk 3 above)
Comment (vvo) [inline lib/style.css]: `.as-hide`? (hidden could be mapped to visibility: hidden)

File: widgets/refinement-list.js (hunk 4 above)
Comment (vvo) [inline widgets/refinement-list.js]: If you did not pass any `templates: {}` then we do not need to merge anything. is that ok?

**Comment by vvo** [comment]
Once we agree on this one on the feature side we can merge it I will submit a new one which will add the same functionnality for every other widget where it makes sense.

And I will need to refactor some things since we are starting to have diplicated code

File: widgets/refinement-list.js (hunk 5 above)
Comment (vvo) [inline widgets/refinement-list.js]: Yes very good idea and implementation :)

Comment (vvo) [inline widgets/refinement-list.js]: As I said this refactor will go in another PR (where I will add the header/footer to every widget)

**Pull request merged**